Security considerations (#873)
mlswg/mls-protocol: draft-ietf-mls-protocol.md

Post-compromise security is provided between epochs by members regularly
updating their leaf key in the ratchet tree. Updating their leaf key prevents
group secrets from continuing to be encrypted to public keys whose private keys

> had previously been compromised. Note that sending an Update does not achieve
> PCS until another member sends a Commit. Members can achieve immediate

Reviewer: ```suggestion
had previously been compromised. Note that sending an Update proposal does not achieve
PCS until another member includes it in a Commit. Members can achieve immediate
```

PCS by sending their own Commit and populating the `path` field, as

> described in {{commit}}.

Reviewer 2: This seems like a worthwhile distinction.  Something like:

> To be clear, in all these cases, the PCS guarantees come into effect when the members of the group process the relevant Commit, not when the sender creates it.

attack, the KeyPackage publication system SHOULD rate-limit KeyPackage
requests, especially if not authenticated.

## Delivery Service Compromise

MLS is designed to protect the confidentiality and integrity of
the group data even in the face of a compromised DS. However, a compromised

> DNS can still mount some attacks. While it cannot forge messages,

Reviewer: ```suggestion
DS can still mount some attacks. While it cannot forge messages,
```

it can selectively delay or remove them. This can in some cases be
observed by detecting gaps in the per-sender generation counter,

> though it may not always be possible to distinguish attack from message

Reviewer: ```suggestion
though it may not always be possible to distinguish an attack from message
```

loss. In addition, the DS can permanently block messages to and from
a group member. This will not always be detectable by other members.
If an application uses the DS to resolve conflicts between
simultaneous Commits (see {{sequencing}}), it is also possible for the
DS to influence which Commit is applied, even to the point of

> preventing a member from ever having its Commit applied.

Reviewer: ```suggestion
preventing a member from ever having its Commits applied.
```

When put together, these abilities potentially allow a DS to collude
with a compromised member to defeat PCS by suppressing the valid
Update and Commit messages sent by the uncompromised endpoint for the

> same user. Aside from the generation, MLS leaves loss detection up the

Reviewer: If I understand correctly, the "compromised member" and "the uncompromised endpoint" are more or less the same here, aren't they? It might make sense to spell the scenario out a bit more.

Also "generation" refers to the generation in the messages, right? Since it's a relatively generic word, it might make sense to be explicit about that context.

Reviewer 2: Agree that this is a little ambiguous.  Suggest something like:

> When put together, these abilities potentially allow a DS to collude with an attacker who has compromised a member's state to defeat PCS by suppressing the valid Update and Commit messages from the member that would lock out the attacker and update the member's leaf to a new, uncompromised state.

> same user. Aside from the generation, MLS leaves loss detection up the

Reviewer 3: ```suggestion
same user. Aside from the generation, MLS leaves loss detection up to the
```

application.

## Authentication Service Compromise

Authentication Service compromise is much more serious than compromise
of the Delivery Service. A compromised AS can assert a binding for a
signature key and identity pair of its choice, thus allowing
impersonation of a given user. This ability is sufficient to allow the
AS to join new groups as if it were that user. Depending on the
application architecture, it may also be sufficient to allow the
compromised AS to join the group as an existing user, for instance as
if it were a new device associated with the same user. If
the application uses a transparency mechanism such as CONIKS
{{CONIKS}} or Key Transparency {{KT}}, then it may be possible for end

> users to detect this kind of misbehavior by the AS.

Reviewer: Since we're naming mitigation measures here, cross-signing of clients would also be worth mentioning here, I think.

Reviewer 2: ```suggestion
users to detect this kind of misbehavior by the AS.  It is also possible to 
construct schemes in which the various clients owned by a user vouch
for each other, e.g., by signing each others' keys.
```

@kkohbrok - If you had a citation, that would be helpful.
